infra/terraform/terraform.go

```diff
@@ -64,45 +71,73 @@ func NewFromState(config Config, stateConfig infra.ProvisionerState) (*terraform
 }
 
 func (r *terraform) Create(withInstaller bool) (installer infra.Node, err error) {
-	err = system.CopyFile(filepath.Join(r.stateDir, "terraform.tf"), r.ScriptPath)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), terraformTimeout)
+	defer cancel()
 
-	output, err := r.boot()
+	err, nfiles := system.CopyAll(r.ScriptPath, r.stateDir)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if nfiles == 0 {
+		return nil, trace.Errorf("No Terraform configs at %s", r.ScriptPath)
+	}
+
+	// sometimes terraform cannot receive all required params
+	// most often public IPs take time to allocate (on Azure)
+	for {
+		err, retry := r.terraform()
+		if err == nil {
+			if withInstaller {
+				nodes := r.pool.Nodes()
+				if len(nodes) == 0 { // should not happen, and doesn't make sense to retry
+					return nil, trace.Errorf("Zero nodes were allocated")
+				}
+				r.installerIP = nodes[0].Addr()
+				return nodes[0], nil
+			}
+			return nil, nil
```

Comment:
Usually, if a function returns `nil` error, the called expects the first value to be usable, so I would avoid returning `nil, nil`. Return an appropriate error instead, like `trace.NotFound()` or `trace.BadParameter()`.